hamster_gtk/misc/widgets/raw_fact_entry.py

```diff
@@ -0,0 +1,229 @@
+# -*- encoding: utf-8 -*-
+
+
+# This file is part of 'hamster-gtk'.
+#
+# 'hamster-gtk' is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# 'hamster-gtk' is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with 'hamster-gtk'.  If not, see <http://www.gnu.org/licenses/>.
+
+"""Widget meant to handle 'raw-fact' strings and provides autocompletion."""
+from __future__ import absolute_import, unicode_literals
+import datetime
+
+from gi.repository import GObject, Gtk
+from orderedset import OrderedSet
+from six import text_type
+from hamster_gtk.helpers import _u
+from hamster_gtk import helpers
+
+
+class RawFactEntry(Gtk.Entry):
+    """A custom entry widgets that provides ``faw fact`` specific autocompletion behaviour."""
+
+    def __init__(self, controller, *args, **kwargs):
+        """Instantiate class."""
+        super(RawFactEntry, self).__init__(*args, **kwargs)
+        self._controller = controller
+        self._controller.signal_handler.connect('facts-changed', self._on_facts_changed)
+        self.set_completion(RawFactCompletion(self._controller))
+        # The re.match instance of the current string or None
+        self.match = None
+        # Identifier for the segment the cursor is currently in. None if no
+        # match is available.
+        self.current_segment = None
+        self.connect('changed', self._on_changed)
+
+    def replace_segment_string(self, segment_string, add_prefix=True):
```

Comment:
I think it is a rather generic method and thought to make it part of the public API. Which is also why I included the ``add_prefix`` in the signature - to increase user ease. But you are right, I can not really come up with a usecase where one would want to deal with prefixing itself (e.g. one would use ``add_prefix=False``). Going to remove it.